holopy/scattering/scatterer/composite.py

```diff
@@ -74,45 +85,71 @@ def get_component_list(self):
                 components.append(s)
         return components
 
-    @property
-    def parameters(self):
-        pars = []
-        names = []
-        for i, scatterer in enumerate(self.scatterers):
-            for key, par in scatterer.parameters.items():
-                for index, par_check in enumerate(pars + [None]):
+    def update_ties(self):
+        for fullkey, par in self.raw_parameters.items():
+            if fullkey not in sum(self.ties.values(), []):
+                # not already in the list of ties, so check if it should be
+                reference_parameters = dict_without(self.parameters, fullkey)
+                for ref_key, ref_par in reference_parameters.items():
                     # can't simply check par in parameters because then two
                     # priors defined separately, but identically will match
                     # whereas this way they are counted as separate objects.
-                    if par is par_check and not isinstance(par, Number):
-                        # par is a prior and already exists in pars
+                    if par is ref_par and not isinstance(par, Number):
+                        if ref_key in self.ties.keys():
+                            self.ties[ref_key].append(fullkey)
+                        else:
+                            subkey = fullkey.split(':', 1)[1]
+                            if subkey not in self.ties.keys():
+                                self.ties[subkey] = [ref_key, fullkey]
+                            else:
+                                self.ties[fullkey] = [ref_key, fullkey]
                         break
-                if par_check is None:
-                    # loop finished - par is not tied.
-                    names.append('{0}:{1}'.format(i,key))
-                    pars.append(par)
-                else:
-                    # loop encountered a break - parameter is tied
-                    names[index] = key
-        return {key:val for key, val in zip(names, pars)}
 
+    @property
+    def raw_parameters(self):
+        parameters = {}
+        for i, scatterer in enumerate(self.scatterers):
+            single_scatterer_parameters = {'{0}:{1}'.format(i, key): val
+                            for key, val in scatterer.parameters.items()}
+            parameters.update(single_scatterer_parameters)
+        return parameters
+
+    @property
+    def parameters(self):
+        parameters = self.raw_parameters
+        for tied_name, raw_names in self.ties.items():
+            for raw_name in raw_names:
+                if raw_name not in parameters.keys():
+                    msg = 'Tied parameter {} not present in raw \
```

Comment:
When does this (and the other) error message get raised in practice? It seems that they are raised if something is not created correctly, i.e. only if there is a bug in the code here. If that's the case I think it's great to have them, but for readability you might want to make them ```assert```s rather than ```ValueError```s. If they are possible to raise by the user, then it would be nice to have some tests (at least so other developers can see what you are checking for and what fail cases are).